viewing_party/party.py

```diff
@@ -1,23 +1,184 @@
+        watched_by_any_friend = False
+
+        # checks if the movie is has been watched by a friend
+        for friend in user_data["friends"]:
+            if movie_watched_by_user in friend["watched"]:
+                watched_by_any_friend = True
+                break
+
+        # If no friend has watched it, adds it to the final list
+        if not watched_by_any_friend:
+            movies_not_watched_by_friends.append(movie_watched_by_user)
+    
+    return movies_not_watched_by_friends
+
+def get_friends_unique_watched(user_data):
 
+    movies_watched_by_friends_not_user = []
+
+    # loops through each friend's watched list
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            # appends to list if 1) movie is not in the user's watched list and 2) not already in the result list
+            if movie not in user_data["watched"] and movie not in movies_watched_by_friends_not_user:
+                movies_watched_by_friends_not_user.append(movie)
+    
+    return movies_watched_by_friends_not_user
         
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+    subscriptions = set(user_data.get("subscriptions", []))
+
+    recommended_movies = set()  # Using a set to ensure uniqueness
+    user_watched = set()  # Track movies the user has already watched
+
+    # Populate user_watched set with movie titles from user_data
+    for movie in user_data.get("watched"):
+        if "title" in movie:
+            user_watched.add(movie["title"])
+  
+    # Loop through each friend and their watched movies
+    for friend in user_data.get("friends"):
+        for movie in friend.get("watched"):
+                # Only add the movie if it's hosted on a subscribed platform and not watched by the user
+            if movie["host"] in subscriptions and movie["title"] not in user_watched:
+                recommended_movies.add((movie["title"], movie["host"], movie["rating"], movie["genre"]))
+
+    # Convert the set of tuples back to a list of dictionaries, including the 'genre' key
+    updated_recommendation = [
+        {"title": title, "host": host, "rating": rating, "genre": genre}
+        for title, host, rating, genre in recommended_movies
+    ]
+
+    return updated_recommendation
+ 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    watched_movies = user_data["watched"]
+    if not watched_movies:
+        return []  # If there are no watched movies, return an empty list
+
+    # Count the genres of watched movies
+    genre_counter = {}
+    for movie in watched_movies:
+        genre = movie["genre"]
+        if genre not in genre_counter:
+            genre_counter[genre] = 0
+        genre_counter[genre] += 1

```

Comment:
This is the frequency map concept I was referring to in [this](https://github.com/Ada-C22/viewing-party/pull/21/files#r1779349599) comment. Nice work, implementing it here!